src/DistributedLocking/Consul/LockSessionProvider.php

```diff
@@ -0,0 +1,117 @@
+<?php
+
+namespace CascadeEnergy\DistributedLocking\Consul;
+
+use CascadeEnergy\DistributedLocking\ILockSessionProvider;
+use CascadeEnergy\Exceptions\ExceptionWithContext;
+use SensioLabs\Consul\Services\Session;
+
+class LockSessionProvider implements ILockSessionProvider
+{
+    /** The default TTL of a session, in seconds */
+    const DEFAULT_TTL = 30;
+
+    /** The minimum number of seconds between outbound heartbeat calls */
+    const HEARTBEAT_INTERVAL = 5;
+
+    private $consulSession;
+
+    /** @var callable A function which returns the current time in (possibly fractional) seconds */
+    private $currentTimeFn;
+
+    private $lastHeartbeatTimestamp = 0;
+    private $sessionId;
+    private $sessionTtl;
+
+    /**
+     * @param Session $consulSession
+     * @param int $ttl The session TTL, in seconds
+     * @param callable $currentTimeFn A function which returns the current time in seconds (defaults to `microtime`)
+     */
+    public function __construct(Session $consulSession, $ttl = self::DEFAULT_TTL, callable $currentTimeFn = null)
```

Comment:
Can you not set `$currentTimeFn` to a function by default? I see where below you're checking `is_null` and then setting it to `function () { return microtime(true); }`. Is it possible in PHP to do:

```php
$microtimeFn = function () {
    return microtime(true);
}

class LockSessionProvider implements ILockSessionProvider
{
// ...
    public function __construct(Session $consulSession, $ttl = self::DEFAULT_TTL, callable $currentTimeFn = $microtimeFn)
    {
        // ....
    }
}
```